src/python/gudhi/reader_utils.cc

```diff
@@ -0,0 +1,103 @@
+/*    This file is part of the Gudhi Library - https://gudhi.inria.fr/ - which is released under MIT.
+ *    See file LICENSE or go to https://gudhi.inria.fr/licensing/ for full license details.
+ *    Author(s):       Vincent Rouvreau
+ *
+ *    Copyright (C) 2020 Inria
+ *
+ *    Modification(s):
+ *      - YYYY/MM Author: Description of the modification
+ */
+
+#include <gudhi/Points_off_io.h>
+#include <gudhi/reader_utils.h>
+
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+#include <pybind11/numpy.h>
+
+#include <vector>
+#include <array>
+#include <string>
+
+namespace py = pybind11;
+
+
+PYBIND11_MODULE(reader_utils, m) {
+      m.attr("__license__") = "MIT";
+      m.def("read_points_from_off_file", [](const std::string& off_file) {
+            Gudhi::Points_off_reader<std::vector<double>> off_reader(off_file);
+            return off_reader.get_point_cloud();
+          },
+        py::arg("off_file"),
+        R"pbdoc(
+    Read points from OFF file.
+
+    :param off_file: An OFF file style name.
+    :type off_file: string
+
+    :returns:  The point set.
+    :rtype: List[List[float]]
+        )pbdoc"
+      );
+      m.def("read_lower_triangular_matrix_from_csv_file", &Gudhi::read_lower_triangular_matrix_from_csv_file<double>,
+        py::arg("csv_file"), py::arg("separator") = static_cast<const char>(';'),
+        R"pbdoc(
+    Read lower triangular matrix from a CSV style file.
+
+    :param csv_file: A CSV file style name.
+    :type csv_file: string
+    :param separator: The value separator in the CSV file. Default value is ';'
+    :type separator: char
+
+    :returns:  The lower triangular matrix.
+    :rtype: List[List[float]]
+        )pbdoc"
+      );
+      m.def("read_persistence_intervals_grouped_by_dimension", &Gudhi::read_persistence_intervals_grouped_by_dimension,
+        py::arg("persistence_file"),
+        R"pbdoc(
+    Reads a file containing persistence intervals.
+    Each line might contain 2, 3 or 4 values: [[field] dimension] birth death
+    The return value is a `dict(dim, list(tuple(birth, death)))`
+    where `dim` is an `int`, `birth` a `float`, and `death` a `float`.
+    Note: the function does not check that birth <= death.
+
+    :param persistence_file: A persistence file style name.
+    :type persistence_file: string
+
+    :returns:  The persistence pairs grouped by dimension.
+    :rtype: Dict[int, List[Tuple[float, float]]]
+        )pbdoc"
+      );
+      m.def("read_persistence_intervals_in_dimension", [](const std::string& persistence_file, int only_this_dim) {
+            // std::vector<std::pair<double, double>> as in read_persistence_intervals_in_dimension
+            // was not giving the good results because of memory alignment
```

Comment:
I wanted to replay my example and it worked... You are right, and I went to fast with this conclusion.

If I go back with a `vector<pair>`, I cannot bind the vector as you proposed.
This is a PyBind11 internal error, but when I try with a `vector<pair>` I have the following error:
```
pybind11/stl_bind.h:377:5: error: static assertion failed: There is not an appropriate format descriptor for this vector
     static_assert(vector_has_data_and_format<Vector>::value, "There is not an appropriate format descriptor for this vector");
     ^~~~~~~~~~~~~
pybind11/stl_bind.h:380:33: error: ‘format’ is not a member of ‘pybind11::format_descriptor<std::pair<double, double>, void>’
     format_descriptor<T>::format();
```
Do I have to bind the pair as well ?

Reply:
Yes, I had gotten the same error. I didn't experiment, so I have no idea what needs tweaking (maybe binders indeed). I would guess that format_descriptor, which is specialized for `T[N]` and `std::array<T,N>`, should also be specialized for double/tuple (or more generally for anything that supports tuple_element/tuple_size), at least when all types are the same. It may be worth opening an issue in pybind11. But maybe that also requires specializing other classes at the same time, or maybe it has undesirable side effects...
(using `std::pair<T,T>` is often considered a bad idea nowadays, as demonstrated by the fact that the committee refused to abbreviate it as `std::pair<T>`)

Reply:
I changed this wrong comment on 8b374238

Reply:
I reported the issue with std::pair as https://github.com/pybind/pybind11/issues/2991 . I also indicated a way we can make it work, if we really need it.